### DISCUSSION TITLE
Use of `install_requires = file: requirements.txt` raises parse error because hashes are used

### DISCUSSION THREAD

**Discussion by phorward** [pypa/setuptools#3433] [Q&A] [answered]
Hello!

I want to use `install_requires = file: requirements.txt` in my setup.cfg to avoid maintaining dependencies twice.

According #1951 and #3253 the `file:` feature for requires is already available since a while, and I'm on setuptools==62, but I'm getting an exception indicating a parse error `pkg_resources.extern.packaging.requirements.InvalidRequirement: Parse error at "'--hash=s'": Expected string_end` because the requirements.txt uses hashes.

Is there any (alternative) way to achieve this problem currently?

**Comment by abravalheri** [comment, marked as answer]
Hi @phorward, this kind of syntax is intentionally not supported.

The `file:` directive in setuptools requires that each line in the file conform with [PEP 508](https://peps.python.org/pep-0508/). It is not intended to be fully compatible with `pip` requirements files.

This is mentioned in the docs[^1] [^2]:

> ... the file format for specifying dependencies resembles a requirements.txt file, however please keep in mind that all non-comment lines must conform with [PEP 508](https://peps.python.org/pep-0508/) (pip-specify syntaxes, e.g. -c/-r/-e flags, are not supported).

Currently in the ecosystem it is not possible to translate such hash requirements to [package metadata](https://packaging.python.org/en/latest/specifications/core-metadata/#requires-dist-multiple-use). So, I don't think there are any fully-compatible alternative for this.

I suppose you can create your own `setup.py` script that reads the file and strips out `pip`-specific syntaxes, but that is up to the developer to maintain (and the semantics would differ... there is simply no straight-forward way to guarantee hashes are respected for dependencies in Python distribution packages)

[^1]: https://setuptools.pypa.io/en/latest/userguide/declarative_config.html#opt-6
[^2]: https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata

**Reply by abravalheri** [reply]
You might have some luck if you use a direct reference URL with hashes as specified in [PEP 440](https://peps.python.org/pep-0440/#direct-references) in your `requirements.txt` file, instead of `--hash`. However you might not be able to upload your package to PyPI (PyPI does not accept direct references).

**Comment by phorward** [comment]
Hello @abravalheri, thanks for the clarification. I think I've now found a way to solve it adequately in setup.py.
